Allow custom `apply` and `unapply` methods in case class companions

adriaanm opened the pull request:
Don't emit a synthetic `apply` (or `unapply`) when it would
clash with an existing one. This allows e.g., a `private apply`,
along with a `case class` with a `private` constructor.

We have to retract the synthetic method in a pretty roundabout way,
as we need the other methods and the owner to be completed already.
Unless we have to complete the synthetic `apply` while completing
the user-defined one, this should not be a problem. If this does
happen, this implies there's a cycle in computing the user-defined
signature and the synthetic one, which is not allowed.

Community build: https://scala-ci.typesafe.com/job/scala-2.11.x-integrate-community-build/541/console

adriaanm commented:
/rebuild

adriaanm commented:
/synch

adriaanm commented:
Community build with forked paradise: https://scala-ci.typesafe.com/job/scala-2.11.x-integrate-community-build/535/console

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/Namers.scala:
```diff
@@ -946,48 +978,63 @@ trait Namers extends MethodSynthesis {
+      //   <synthetic> <case> def unapply[Ts](x: C[Ts]) = <ret-val>
+      //  where <ret-val> is the caseClassUnapplyReturnValue of class C (see UnApplies.scala)
+      //
+      // This last condition requires special care, as we need to detect overloading without forcing too
```
TODO: finish my sentence

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Namers.scala:
```diff
@@ -463,31 +434,35 @@ trait Namers extends MethodSynthesis {
       }
       else {
-        m = assignAndEnterSymbol(tree)
+        enterInScope(assignMemberSymbol(tree))
```
This is another candidate for "enterInScope may return existing symbol? In interactive setting?"

If this happens (`enterInScope` returns a different symbol), it looks like `tree.symbol` no longer holds the correct symbol. Not sure if this causes issues. In any case, this commit doesn't make things worse.

lrytz commented on test/files/neg/userdefined_apply.scala:
```diff
@@ -0,0 +1,33 @@
+}
+case class ClashOverloadNoSig private (x: Int)
+
+// needs sig because it's recursive
```
error message would ideally also be more clear in this case ("recursive method needs result type"), but the cure is the same.

lrytz reviewed (commented):
Looks very good!

There's an empty commit (676ec1618a7e99c1874ba8cbb6b09fd0acaf1c77).

adriaanm commented:
I'll see about moving some of the more ambitious changes over to 2.12.2. Is there anything before 		09c35ae (incl) that you're concerned about?

adriaanm commented:
On second thought, I'll just make the minimal changes here and move the cleanups to 2.12

lrytz commented:
It LGTM, but agree it's a good idea to make minimal changes in 2.11.

adriaanm committed:
Allow user-defined `[un]apply` in case companion

Don't emit a synthetic `apply` (or `unapply`) when it would
clash with an existing one. This allows e.g., a `private apply`,
along with a `case class` with a `private` constructor.

We have to retract the synthetic method in a pretty roundabout way,
as we need the other methods and the owner to be completed already.
Unless we have to complete the synthetic `apply` while completing
the user-defined one, this should not be a problem. If this does
happen, this implies there's a cycle in computing the user-defined
signature and the synthetic one, which is not allowed.

adriaanm commented:
Ready for final 👀 .

adriaanm commented:
More ambitious wip for 2.12 moved to https://github.com/scala/scala/compare/2.12.x...adriaanm:userdefined-apply-212?expand=1

SethTisue commented:
Does this need a spec update?

adriaanm commented:
True:
  - [ ] If a companion object cc exists already, no new object is created, but the apply and unapply methods are added to the existing object instead, unless the object already has a [matching](#definition-matching) definition.

adriaanm committed:
Clarify spec of interaction of existing vs synthetic apply/unapply

When matching user-defined apply/unapply members exist in a
case class's companion object, don't add clashing synthetic ones.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Namers.scala:
```diff
@@ -629,14 +637,57 @@ trait Namers extends MethodSynthesis {
+
+          // If there's a same-named locked symbol, we're currently completing its signature.
+          // This means it (may) refer to us, and is thus either overloaded or recursive without a signature.
+          // rule out locked symbols from the owner.info.member call
```
Tricky stuff.. so if there's a locked sym in the same scope, the current sym is marked error. Does that cause the error to be reported? Why do we want to avoid the `owner.info.member` member call in this case?

Could there be a similar situation when there's a locked symbol not in the same scope, but inherited?

adriaanm replied:
`member` will need the `info` of the symbols its collecting in an `OverloadedType` (to see if they match). Calling `info` on those locked symbols will yield a `TypeError` to indicate the cycle, causing our stack to be unwound.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Namers.scala:
```diff
@@ -629,14 +637,57 @@ trait Namers extends MethodSynthesis {
+            sym setInfo ErrorType
+            sym setFlag IS_ERROR
+
+            // Don't unlink in an error situation to generate less confusing error messages.
```
It's not obvious to me why/how this impacts error reporting..

adriaanm replied:
Unlinking when there actually isn't a clash would result in the errors talking about recursive methods (since `symWasOverloaded` is `false`) instead of overloaded methods, flipping the wrongness of the error message to the other side.
```
    private def symWasOverloaded(sym: Symbol) = sym.owner.isClass && sym.owner.info.member(sym.name).isOverloaded
```

adriaanm replied:
If we could be more precise in determining whether there is a clash (by partially completing the method info in methodSig with a wildcard for the result type when it's inferred), we could make the error messages more accurate. Looking into that for 2.12

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Namers.scala:
```diff
@@ -697,11 +748,17 @@ trait Namers extends MethodSynthesis {
-        else
-          sym setInfo completerOf(tree)
-    }
+        // copy/apply/unapply synthetics are added using the addIfMissing mechanism,
```
`addIfMissing` is no longer in this PR

adriaanm replied:
good catch

lrytz commented on test/files/neg/userdefined_apply.scala:
```diff
@@ -0,0 +1,31 @@
+case class ClashOverloadNoSig private(x: Int)
+
+object ClashRecNoSig {
+  // error: recursive method apply needs result type
```
The actual report says `overloaded method apply needs result type`, as you explain in the lengthy comment in Namers. So the behavior is fine, just the comment here is out of synch.

adriaanm replied:
fixed

lrytz reviewed (commented):
Looks good, just a few clarifications.

fommil commented:
I really like this, it's stung me a couple of times in the past.

fommil commented:
btw, somewhat related, a GSoC proposal I'm putting together for this year: https://github.com/scala/scala-lang/blob/master/gsoc/2017.md#case-classes-a-la-carte-with-scalameta

adriaanm committed:
Improvements based on reviews by Lukas & Jason

retronym commented on spec/05-classes-and-objects.md:
```diff
@@ -870,11 +869,14 @@ each $\mathit{xs}\_i$ denotes the parameter names of the parameter
+The definition of `apply` is omitted if class $c$ is `abstract`.
+If the object $c$ already defines a [matching](#definition-matching) member of the
+same name as the synthetic member to be added, the synthetic member
+is not added (overloading or mutual recursion is allowed, however).
```
"defines or inherits" ?

"mutual recursion" sounds to me like the synthetic apply might call the user-defined apply, whereas it only calls the constructor. Can we tighten up this description, or maybe just avoid speccing the fact that we allow the user-defined apply to call the synthetic one.

adriaanm replied:
fixed

retronym commented on src/compiler/scala/tools/nsc/typechecker/Namers.scala:
```diff
@@ -629,13 +637,58 @@ trait Namers extends MethodSynthesis {
+                case tp =>
+                  (tp eq ErrorType) || tp.matches(sym.info)
+              }
+            }
```
Perhaps something along these lines might read more cleanly?

```
scala> case class C(a: Int); object C { def apply(a: Int, b: Int) = 42; object apply }
defined class C
defined object C

scala> val synthApply = typeOf[C.type].decl(nme.apply).alternatives.find(_.isSynthetic).get // just `sym` in code above
synthApply: $r.intp.global.Symbol = method apply

scala> val userDefinedApplies = typeOf[C.type].memberBasedOnName(nme.apply, Flags.BridgeFlags | Flags.SYNTHETIC).alternatives
userDefinedApplies: List[$r.intp.global.Symbol] = List(object apply, method apply)                                                                       

scala> userDefinedApplies.map(x => (x, typeOf[C.type].memberInfo(x), typeOf[C.type].memberInfo(x) matches synthApply.info)).mkString("\n")
res20: String =
(object apply,C.apply.type,false)
(method apply,(a: Int, b: Int)Int,false)
```

retronym replied:
As we vaguely discussed, it would be create to have less opaque type completers for method symbols that let us find out the parameter list arity first (which could filter out some members already), then the method type with a wildcard result to represent "not yet inferred", and then finally the full method type. It would still be a bit of a pain to use, as we couldn't just reuse `findMember` as it, we'd need to generalise it to allow us to specify a signature schema to narrow the search.

adriaanm replied:
I played with various approaches to partial sigs. Would be really nice, and perhaps feasible in 2.12, but leaving out of scope here.

Could you elaborate on your first comment? Do you intend to avoid the `.filter` by excluding `SYNTHETIC`?

retronym replied:
If we don't exclude the `SYNTHETIC` member in from `findMember`, I don't think we'll notice if the object would have inherited a matching member. (We'd need to separately check the override chain of the synthetic one.) Excluding it from `findMember` seems to kill two birds with one stone.

adriaanm replied:
Ah yes. Hmm, I wonder how my pos test didn't catch this.

adriaanm replied:
Also had to exclude DEFERRED, since otherwise we suppress our synthetic implementation of the apply inherited from FunctionN (as it's a matching member, albeit not a concrete one)

retronym commented on src/compiler/scala/tools/nsc/typechecker/Namers.scala:
```diff
@@ -629,13 +637,58 @@ trait Namers extends MethodSynthesis {
+          // If there's a same-named locked symbol, we're currently completing its signature.
+          // This means it (may) refer to us, and is thus either overloaded or recursive without a signature.
+          val scopePartiallyCompleted =
+            companionContext.scope.lookupAll(sym.name).exists(existing => existing != sym && existing.hasFlag(LOCKED))
```
This won't find cycles involving inherited `apply`-s.

adriaanm replied:
Good point! Added a `HasMember` subclass of `FindMemberBase` for a more complete check (which does not force any infos). Also added test.

retronym commented on src/compiler/scala/tools/nsc/typechecker/Namers.scala:
```diff
@@ -629,13 +637,58 @@ trait Namers extends MethodSynthesis {
+            // as FindMember will call info on a locked symbol (while checking type matching to assemble an overloaded type),
+            // and throw a TypeError, so that we are aborted.
+            scopePartiallyCompleted || {
+            val userDefined = companionContext.owner.info.member(sym.name).filter(_ != sym)
```
Is there a chance that `companionContext.owner` is locked at this point?

retronym replied:
Maybe not. I was thinking about bugs like:

https://issues.scala-lang.org/browse/SI-6357

But here, `companionContext.owner` is the just the module class symbol, not its enclosing owner.

retronym commented on src/compiler/scala/tools/nsc/typechecker/Namers.scala:
```diff
@@ -629,13 +637,58 @@ trait Namers extends MethodSynthesis {
+                  // pre probably relevant because of inherited overloads?
+                  alternatives.exists(_.isErroneous) || alternatives.exists(alt => pre.memberInfo(alt) matches pre.memberInfo(sym))
+                case tp =>
+                  (tp eq ErrorType) || tp.matches(sym.info)
```
This line neglects the `memberInfo` on the (potentially inherited) user defined member.

adriaanm replied:
ASF strikes again. Fixed, and test added.

lrytz commented on src/reflect/scala/reflect/internal/tpe/FindMembers.scala:
```diff
@@ -285,4 +285,21 @@ trait FindMembers {
     }
   }
+
+  private[scala] final class HasMember(tpe: Type, name: Name, excludedFlags: Long, requiredFlags: Long, stableOnly: Boolean)
```
`stableOnly` is not used

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Namers.scala:
```diff
@@ -649,23 +649,34 @@ trait Namers extends MethodSynthesis {
+          // The method would be overloaded if its signature, once completed, would not match the synthetic method's,
+          // or recursive if it turned out we should unlink our synthetic method (matching sig).
+          // In any case, error out. We don't unlink the symbol so that `symWasOverloaded` says yes,
+          // which would be wrong if the method is in fact recursive, but it seems less confusing.
```
thanks, this much more clear now 👍

retronym commented on src/reflect/scala/reflect/internal/tpe/FindMembers.scala:
```diff
@@ -285,4 +285,21 @@ trait FindMembers {
+
+    protected def shortCircuit(sym: Symbol): Boolean = {
+      _result = sym
+      true // prevents call to addMemberIfNew
```
AFAICT this will only find the first member with `name`, which doesn't seem like what we want: we need to consider all members.

It would be possible to change `addMemberIfNew` to just to treat everything as overrides (which would basically devolve to `cls.baseClasses.flatMap(_.decl(name)).filter(x => !x.isPrivate || x.owner == cls)`. That might avoids the `lo matches hi` checks needs to distinguish overrides from overloads.

We still need to force the infos of each member to see if they matches the synthetic method type, so I'm not sure whether it is any lazier on type completion.

adriaanm replied:
In this case, all we care about finding any member of said name with the locked flag.
I guess we should return a boolean to avoid confusion.

retronym commented:
Maybe it flukes it with a fortuitous order of members?

adriaanm commented:
Well then, looks like this should be the one 🤞

retronym commented:
When I reviewed the last iteration of this change, I mistakenly thought that the custom `FindMember` implementation replaced the entire call to `.members`, rather than just the search for the existence of a locked decl. That's why I thought it was a mistake to short circuit.

This is the code I'd imagined at the time that joins these two parts into a single member walk:

https://github.com/retronym/scala/tree/review/5730

It also avoids completing infos of members over trees that don't match the arity.

That said, I'm happy enough with this change as it is. Feel free to take some ideas from my branch if you further rework this on the 2.12.x branch, but I'm giving this the 👍 here.

adriaanm commented:
Thanks, for the next refinement, I'd prefer to enrich the type completer interface so that you can ask more directly for partial completions (e.g., just the types you can determine without looking at RHS, just the type params, ...)

retronym commented:
Sounds good. e.g. we could add a method like `schematicType` that gives us the shape of the method type with wildcards. To be useful, we've have to change a few other places to use them, notable the checks for overriden members in namers, such as:

```
      val overridden = {
        val isConstr   = meth.isConstructor
        if (isConstr || !methOwner.isClass) NoSymbol else overriddenSymbol(methResTp)
      }
      val hasDefaults = mexists(vparamss)(_.symbol.hasDefault) || mexists(overridden.paramss)(_.hasDefault)
      if (hasDefaults)
        addDefaultGetters(meth, ddef, vparamss, tparams, overridden)
```

adriaanm commented:
Right, having a way to trigger cycle-free completion (peeling off just one
layer) may also enable experimenting with more pro-active completion
schemes so that we can populate the symbol table in a tighter loop, without
jumping back and forth and going into deep lazy completer call chains.
On Thu, Mar 2, 2017 at 16:45 Jason Zaugg <notifications@github.com> wrote:

> Sounds good. e.g. we could add a method like schematicType that gives us
> the shape of the method type with wildcards. To be useful, we've have to
> change a few other places to use them, notable the checks for overriden
> members in namers, such as:
>
>       val overridden = {
>         val isConstr   = meth.isConstructor
>         if (isConstr || !methOwner.isClass) NoSymbol else overriddenSymbol(methResTp)
>       }
>       val hasDefaults = mexists(vparamss)(_.symbol.hasDefault) || mexists(overridden.paramss)(_.hasDefault)
>       if (hasDefaults)
>         addDefaultGetters(meth, ddef, vparamss, tparams, overridden)
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/scala/scala/pull/5730#issuecomment-283830304>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAFjy10wqMr1jhx3X0VYRqg-yB0tfp1Wks5rh2KWgaJpZM4MJozQ>
> .
>
